Feature/5114 add parallel commit config

mathbraga opened the pull request:
## :bookmark_tabs: Summary

Adds support for parallel commits in git graphs.

Resolves #5114 

## :straight_ruler: Design Decisions

The logic consists on searching for the closest parent commit of the current commit being positioned and placing it right after said parent, instead of the default that places after the latest commit.

`parallelCommits: false`
![image](https://github.com/mermaid-js/mermaid/assets/31048764/2716fae7-36f8-4457-aa3f-20e4ce9849c5)

`parallelCommits: true`
![image](https://github.com/mermaid-js/mermaid/assets/31048764/decdb701-4358-4281-9703-ebec2d0871f7)

### :clipboard: Tasks

Make sure you

- [x] :book: have read the [contribution guidelines](https://github.com/mermaid-js/mermaid/blob/develop/CONTRIBUTING.md)
- [x] :computer: have added necessary unit/e2e tests.
- [ ] :notebook: have added documentation. Make sure [`MERMAID_RELEASE_VERSION`](https://github.com/mermaid-js/mermaid/blob/develop/packages/mermaid/src/docs/community/development.md#3-update-documentation) is used for all new features.
- [x] :bookmark: targeted `develop` branch

nirname commented on packages/mermaid/src/diagrams/git/gitGraphRenderer.js:
```diff
@@ -87,9 +110,16 @@ const drawCommits = (svg, commits, modifyGraph) => {
 
+    if (isParallelCommits && commit.parents.length) {
+      const closestParent = findClosestParent(commit.parents, commits);
+      pos = dir === 'TB' ? commitPos[closestParent].y + 40 : commitPos[closestParent].x + 40;
```
There is no check for the case when `closesParent` is blank string. I am not sure, maybe there is always an empty element in `commitPos`, but it seems strange a little bit. Initial commit should not have a parent, for instance, if I am not mistaken

mathbraga replied:
`closestParent` will never be blank here I believe. The `commit.parents.length` in the `if` condition garantees that the commit has atleast 1 parent, so the `findClosestParent` inside always returns a parent. Initial commit for example won't satisfy this `if` condition since its `commit.parents.length` will be 0.

nirname commented on packages/mermaid/src/diagrams/git/gitGraphRenderer.js:
```diff
@@ -64,6 +64,29 @@ const drawText = (txt) => {
+ * @returns {string}
+ */
+const findClosestParent = (parents) => {
+  let closestParent = '';
```
Maybe it is better to return undefined when parent is missing

mathbraga replied:
The way I used `findClosestParent` in the code guarantees the function returns a parent as I mentioned above. But I do agree to return `undefined` for correctness, and the function could be used in other contexts aswell.

nirname reviewed (changes requested):
This example shows that parallel commits do not work in some cases
![image](https://github.com/mermaid-js/mermaid/assets/2842432/6513e23b-4a62-4f09-b957-c5726f648c1a)

```
---
config:
  gitGraph:
    parallelCommits: true
---
gitGraph
branch dev
branch v2
branch feat
commit
commit
checkout main
commit
checkout dev
commit
checkout v2
commit
checkout main
commit
```

There is a [link to check this example](https://deploy-preview-5161--mermaid-js.netlify.app/edit#pako:eNp9UDEOwjAM_ErkuV0Ys4LExMSaxSRuG5E4VeogIcTfSYpASCAmn0_ns-5uYJMj0ND3vWGbePCjNqzU6GWfcZ7WRakZM4ZAYZti9LJoJbmQ4fXqJTV8ysh2Uo4ub3zZvOFAKO1Hc_iYE9lzKqIiev5Br2ZfbLP96wAdRMqVcTXcrYUwIBNFMqArdDRgCWLA8L1KsUg6XtmCbrk6KLNDoZ3HMWMEPWBYKkvOS8qHZ2Frb_cHNzh0hg)

nirname commented:
Great feature! Some minor changes are needed.

mathbraga commented:
Made some of the changes you suggested. Please confirm if you still want me to change the logic with the if condition after reading the explanation at [the comment above](https://github.com/mermaid-js/mermaid/pull/5161#discussion_r1454164599).

I fixed the issue from your example and added some e2e tests using it.

sidharthv96 commented on packages/mermaid/src/diagrams/git/gitGraphRenderer.js:
```diff
@@ -87,9 +110,23 @@ const drawCommits = (svg, commits, modifyGraph) => {
+      if (!commit.parents.length) {
+        pos = 0;
+        if (dir === 'TB') {
+          pos = 30;
+        }
+      } else {
+        const closestParent = findClosestParent(commit.parents);
+        pos = dir === 'TB' ? commitPos[closestParent].y + 40 : commitPos[closestParent].x + 40;
```
If we move the magic numbers 40 and 10 to constants, that would be nice.

mathbraga replied:
Noticed the `pos + 10` was very repetitive throughout the code so I also made that into a const. Some of the 10 values seemed to have a different context so I left them untouched.

sidharthv96 reviewed (approved):
Just a minor comment (optional).

sidharthv96 commented on packages/mermaid/src/diagrams/git/gitGraphRenderer.js:
```diff
@@ -330,7 +373,7 @@ const drawCommits = (svg, commits, modifyGraph) => {
         }
       }
     }
-    pos += 50;
+    pos += commitStep + layoutOffset;
```
❤️

sidharthv96 commented on packages/mermaid/src/diagrams/git/gitGraphRenderer.js:
```diff
@@ -87,11 +110,31 @@ const drawCommits = (svg, commits, modifyGraph) => {
+  const commitStep = 40;
   sortedKeys.forEach((key) => {
     const commit = commits[key];
+    const posWithOffset = pos + layoutOffset;
+
+    if (isParallelCommits) {
+      if (!commit.parents.length) {
+        pos = 0;
+        if (dir === 'TB') {
+          pos = 30;
+        }
+      } else {
+        const closestParent = findClosestParent(commit.parents);
+        pos =
+          dir === 'TB'
+            ? commitPos[closestParent].y + commitStep
+            : commitPos[closestParent].x + commitStep;
+      }
+    }
```
Shouldn't posWithOffset be set after the if block, as pos is updated inside it?

mathbraga replied:
Yes it should, it even breaks the tests I added, not sure how I missed this. Good catch.

sidharthv96 commented on packages/mermaid/src/diagrams/git/gitGraphRenderer.js:
```diff
@@ -87,11 +110,31 @@ const drawCommits = (svg, commits, modifyGraph) => {
+    const posWithOffset = pos + layoutOffset;
+
+    if (isParallelCommits) {
+      if (!commit.parents.length) {
+        pos = 0;
+        if (dir === 'TB') {
+          pos = 30;
+        }
+      } else {
+        const closestParent = findClosestParent(commit.parents);
+        pos =
+          dir === 'TB'
+            ? commitPos[closestParent].y + commitStep
+            : commitPos[closestParent].x + commitStep;
+      }
```
Oh, and please swap the blocks of this `if` to avoid the negation.
